### ISSUE TITLE
AddRule/DeleteRule/AddRule for rule session is not working

### ISSUE THREAD

**Issue by magallardo** [project-flogo/rules#80]
**Current behavior (how does the issue manifest):**

1) Adding a rule  the first time works. The rule conditions and actions are executed

2) Deleting the rule works.  Rule conditions and actions are not executed any more

3) Adding the same rule again, it return success but the rule conditions or actions are not executed anymore

4) Adding a complete new rule (with new name) does not work either

**Expected behavior:**
1) Adding a rule  the first time works. The rule conditions and actions are executed

2) Deleting the rule works.  Rule conditions and actions are not executed any more

3) Adding the same rule again, it return success and the rule conditions and actions should be executed

4) Adding a complete new rule (with new name) should also work and conditions and actions should be executed.

**Minimal steps to reproduce the problem (not required if feature enhancement):**
Follow steps 1 to 4 as indicated above

**Please tell us about your environment (Operating system, docker version, browser & web ui version, etc):**
MacOS.

**Flogo Rules version (CLI & contrib/lib. If unknown, leave empty or state unknown):** 0.X.X
Latest version in github.

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue happens all the time.

**Comment by balamg** [comment]
I could not reproduce this behavior. Can you attach a test case that does not work?

**Comment by rameshpolishetti** [comment]
@magallardo I too failed to reproduce the issue assuming that you would have tried adding/deleting rules like below. Let me know if you are trying anything differently:

```code

func TestGitIssue80(t *testing.T) {
	rs, _ := createRuleSession()

	// create rules
	r1 := ruleapi.NewRule("Rule1")
	r1.AddCondition("r1c1", []string{"t1.none"}, trueCondition, nil)
	r1.SetAction(r1Action)
	r2 := ruleapi.NewRule("Rule2")
	r2.AddCondition("r2c1", []string{"t1.none"}, trueCondition, nil)
	r2.SetAction(r2Action)

	// add rule 1
	err := rs.AddRule(r1)
	if err != nil {
		t.Fatalf("not able to add Rule[%s]", r1.GetName())
	}
	t.Logf("Rule[%s] added. \n", r1.GetName())

	// remove rule 1
	rs.DeleteRule(r1.GetName())
	t.Logf("Rule[%s] deleted. \n", r1.GetName())

	// add rule 1 again
	err = rs.AddRule(r1)
	if err != nil {
		t.Fatalf("not able to add Rule[%s] again", r1.GetName())
	}
	t.Logf("Rule[%s] added again. \n", r1.GetName())

	// add rule 2
	err = rs.AddRule(r2)
	if err != nil {
		t.Fatalf("not able to add Rule[%s]", r1.GetName())
	}
	t.Logf("Rule[%s] added. \n", r2.GetName())

        // start session
	rs.Start(nil)

	t1, _ := model.NewTupleWithKeyValues("t1", "t10")
	rs.Assert(context.TODO(), t1)

	rs.Unregister()
}

func r1Action(ctx context.Context, rs model.RuleSession, ruleName string, tuples map[model.TupleType]model.Tuple, ruleCtx model.RuleContext) {
	fmt.Println("r1Action called")
}

func r2Action(ctx context.Context, rs model.RuleSession, ruleName string, tuples map[model.TupleType]model.Tuple, ruleCtx model.RuleContext) {
	fmt.Println("r2Action called")
}

```

**Comment by magallardo** [comment]
@rameshpolishetti I am doing the same to create the rules and delete the rules.  The only difference is that I call a function to add the rule (which has part of the code above), then the rule processes events (conditions and actions get triggered) and then while the events are still being asserted into the engine, I call the function to remove the rule.  That all works fine.
Then the engine continues receiving events, then I call the function to add the rule again, and the conditions/actions of the newly added rule are not evaluated any more.

Please let me know if you have any questions.

**Comment by rameshpolishetti** [comment]
@magallardo Looks like you editing rules (add/delete) even after some events are asserted into the engine.
The current state of the rule engine anticipates all rules added to it before start receiving events. 

@balamg correct me if anything is incorrect.

**Comment by balamg** [comment]
@magallardo @rameshpolishetti It could be a race condition between asserting and creating/deleting rules. I have created branch `issue80` that tries to address it. Can you use this branch and check if it resolves the issue?

**Comment by balamg** [comment]
When a rule is removed and is added back again, it loses its state with respect to the previously asserted tuples. So you need to rebuild that state by bootstrapping the rule with previously asserted tuples
#84 address this

**Comment by balamg** [comment]
closed in #84
